### DISCUSSION TITLE
match unlimited unordered group

### DISCUSSION THREAD

**Discussion by Yokozuna59** [eclipse-langium/langium#1066] [Q&A] [answered]
If I have the following information: `age`, `name`, and `address` (Don't worry about the parse rule itself). Which occurs at a specific location in the language, let's say at the top of the file. For example:

```
age ...
name ...
address ...
```

There is no specific order required. `age` can occurs before and after `name` and `address` and so on.

Each one of those information can occur more than once, but only the last value will be recorded. For example, here `age` value will be `30`:

```
age 20
age 30
```

But how can I parse unordered info that could occurs more than once unlimited times?

For example:

```
name ...
age ...
address ...
age ...
```

---

I've created a `fragment` that can parse repeated information of one type. For example, `age` only or `name` only.

```langium
fragment INFORMATIONS:
    age=AGE+ & name=NAME+ & address=ADDRESS+
;
```

Can parse:

```
age 20
age 30
age 40
```

**Comment by msujew** [comment, marked as answer]
The whole idea behind unordered groups is that they only parse an element once, never twice. Using unordered groups for this doesn't really make sense. You should just use alternatives instead:

```
fragment INFORMATIONS:
    (age=AGE | name=NAME | address=ADDRESS)*
;
```

**Reply by Yokozuna59** [reply]
Oh, I wasn't aware of that, but it didn't work.

Since I've got custom terminal rules for parsing, `age`, `name`, and `address`, I've overridden `TokenBuilder` and changed the `PATTERN` of each of those terminal rules. However, only one of those information it value has been set, leaving the others `undefined`. The others weren't even called to set their values.

**Reply by msujew** [reply]
I don't really follow. What are you trying to do and what is the result/expectation?

**Reply by Yokozuna59** [reply]
I've some placeholder terminal rules:

```langium
terminal AGE: /age/;
terminal NAME: /name/;
terminal ADDRESS: /address/;
```

So, I extended `DefaultTokenBuilder` to override those rules:

```typescript
export class CustomTokenBuilder extends DefaultTokenBuilder {
    override buildTerminalToken(terminal: TerminalRule): TokenType {
        const tokenType = super.buildTerminalToken(terminal);
        switch (tokenType.name) {
            case 'AGE': {
                tokenType.LINE_BREAKS = false;
                tokenType.PATTERN = matchAge;
                break;
            }
            case 'NAME': {
                tokenType.LINE_BREAKS = false;
                tokenType.PATTERN = matchName;
                break;
            }
            // ...etc
        }
        return tokenType;
    }
}
```

When having a single piece of information (one of `age`, `name`, or `address`), it will run through all custom matchers. For example, the following block of code will call all `matchAge`, `matchName`, etc. And only will find one `age`.

```
age 20
```

But when having multiple pieces of information (two or more of `age`, `name`, and `address`), it will only call one of the matchers, although they all exist. For example, the following block of code will only set the value of `name` and drop other information to be `undefined`:

```
age 20
name test
address moon
```

Because `matchAge` and `matchAddress` haven't even been called.

**Reply by msujew** [reply]
It's likely an issue with how these matcher functions are defined. Can you share them? Or preferably a whole reproducible example repo.

**Reply by Yokozuna59** [reply]
I'm not really sure if they are the cause since they weren't even called.

Although they weren't stated above as `age`, `name`, and `address`, here they are inside `title.ts`, `acctitle.ts`, `accdescr.ts`

<https://github.com/Yokozuna59/mermaid-lint/tree/master/packages/mermaid-parser/src/language/matchers>

And here is a test example code:

<https://github.com/Yokozuna59/mermaid-lint/blob/f91cba9bff09ca3c4e907fac2520e15b392692dd/packages/mermaid-parser/tests/common/all.test.ts#L10_L23>

> **Note**: If the code was too complicated please tell to create a replicate repo for the example itself.

**Reply by Yokozuna59** [reply]
And I was wondering, can I perform a custom not global match pattern?

In `chevrotain`, it's possible to create tokens and `CONSUME` them in the order I want. But the `PATTERN` in `langium` uses regex to parse the whole code and might match unwanted patterns. If I have to restrict and handle each case manually, it will increase the complexity and lower the performance of the regex.

**Reply by msujew** [reply]
Looking at your code, it seems like you're using the pattern matching functions incorrectly. Please take a look at [Chevrotains documentation](https://chevrotain.io/docs/guide/custom_token_patterns.html#usage).

These functions accept an offset as their second argument and should only return a token if the text at the offset matches the specific token type. Your code disregards that argument completely, which results in unexpected tokens in the token stream.

Also note that exclusion of other token types is therefore not necessary, as they are tested first. If a token type that appears earlier matches the text at the offset, no other token matching will be attempted.

**Reply by Yokozuna59** [reply]
The example in the chevrotain documentation wasn't complex, which is why I thought of removing the `offset` argument. Then it appears that I need to make major changes to those matching functions according to the new grammar. 

Thank you so much for your help!

**Reply by Yokozuna59** [reply]
One more thing: how can I create a custom hidden rule? Since matching functions should return tokens, hidden rules should remove stuff from the entire text.

**Reply by msujew** [reply]
You should usually just use normal hidden terminals for that. They are just normal Chevrotain token types with the `group` property set to `hidden`. Langium has no concept of hidden rules. If you don't need the result of a rule, you can just assign it to a variable you don't need.

**Reply by Yokozuna59** [reply]
Continuing from this [conversation](https://github.com/Chevrotain/chevrotain/issues/414#issuecomment-1567571942), I modified the matching function to be:

```typescript
const titleRegex = /title(?:[\t ]+([^\n]+)?|$)/y;
export const matchTitle: CustomPatternMatcherFunc = (
    text: string,
    startOffset: number,
) => {
    titleRegex.lastIndex = startOffset;
    let match = titleRegex.exec(text);
    if (match !== null && match[1] !== undefined) {
        match.payload = match[1].trim();
    } else {
        match = null;
    }
    return match;
};
```


But it's appear that `payload` has no effect on the output, from [chevrotain documentation](https://chevrotain.io/docs/guide/custom_token_patterns.html#custom-payloads):

> Sometimes we want to collect additional properties on an IToken object, for example:
> 
> - Save RegExp capturing groups on the token object.
> - Subsets of the matched text, e.g: strip away the quotes from string literals.
> - ...
>
> This can be done by attaching a payload property to our custom token matcher returned value...


So assuming that `title sample   ` as passed, the `parseResult` would has `title sample   ` as it is with out the modification, instead of `sample` only.

**Reply by Yokozuna59** [reply]
It appears that the problem is on Langium's side. I tracked the stack and found the line that caused the unexpected behavior:

<https://github.com/langium/langium/blob/9dcd2416da265f488f84b369b0c3b8f134513a11/packages/langium/src/parser/langium-parser.ts#L195>

There is no validation to determine if `token.payload` exists or not and use it if it does instead of `token.image`. And it's appear the whole method doesn't validate.

I have created a PR here, #1067, that resolves my issue. Although I didn't resolve that issue where the payload might be an object rather than just a string since I'm not sure what the best approach is..
